File: backend/src/main/java/dz/tazrout/dashboard/mqtt/MqttTopics.java
/*
 * FILE: MqttTopics.java
 * PURPOSE: Central registry of all MQTT topic string constants used
 *          throughout the backend. Prevents hardcoded topic strings
 *          in publishers/subscribers. See docs/MQTT_TOPICS.md for
 *          full topic documentation.
 *
 *          Convention: tazrout/{layer}/{zone_or_entity}/{action}
 *
 *          Topics to define:
 *
 *          --- Zone Topics (field ↔ backend ↔ dashboard) ---
 *            tazrout/zones/{zoneId}/sensors          ← gateway publishes sensor readings
 *            tazrout/zones/{zoneId}/state             ← gateway publishes device state (retained)
 *            tazrout/zones/{zoneId}/valve/command      ← AI Engine publishes valve commands
 *            tazrout/zones/{zoneId}/valve/ack          ← gateway publishes command ACK
 *            tazrout/zones/{zoneId}/alert              ← gateway or AI on zone-level anomaly
 *            tazrout/zones/all/sensors                 ← aggregated sensor feed (optional)
 *
 *          --- AI Topics ---
 *            tazrout/ai/decisions                     ← AI Engine publishes all decisions
 *            tazrout/ai/decisions/latest               ← AI Engine publishes latest (retained)
 *            tazrout/ai/alerts                         ← AI Engine publishes critical alerts
 *
 *          --- System Topics ---
 *            tazrout/system/gateway/heartbeat          ← gateway heartbeat (QoS 0)
 *            tazrout/system/gateway/status             ← gateway state (retained)
 *            tazrout/system/backend/status             ← backend status (retained, LWT)
 *            tazrout/system/control                    ← dashboard → backend commands
 *            tazrout/system/emergency/alert            ← AI/gateway critical events (QoS 2)
 *            tazrout/system/emergency/stop             ← dashboard → forces all valves closed
 *            tazrout/system/emergency/status           ← backend → dashboard (retained)
 *
 *          --- Dashboard Topics ---
 *            tazrout/dashboard/summary                ← backend aggregated summary
 *            tazrout/dashboard/performance/{period}    ← backend performance metrics
 *            tazrout/dashboard/notifications           ← backend push notifications
 *
 *          --- Settings & Support ---
 *            tazrout/settings/preferences             ← bidirectional preference sync
 *            tazrout/support/contact                  ← backend → dashboard contact info
 *
 *          --- Analytics ---
 *            tazrout/analytics/water-usage/{period}
 *            tazrout/analytics/environmental/{period}
 *            tazrout/analytics/consumption/{period}
 *
 * MQTT TOPICS: All topics listed above (constants only — no pub/sub logic)
 * DATABASE: N/A
 * DEPENDENCIES: None
 * IMPLEMENTED BY: Mr. Fehis
 */
package dz.tazrout.dashboard.mqtt;

public final class MqttTopics {

    private MqttTopics() {
    }

    public static final String ROOT_WILDCARD = "tazrout/#";

    public static final String ZONE_SENSORS_SUFFIX = "/sensors";
    public static final String ZONE_STATE_SUFFIX = "/state";
    public static final String ZONE_VALVE_ACK_SUFFIX = "/valve/ack";

    public static final String AI_DECISIONS = "tazrout/ai/decisions";
    public static final String AI_DECISIONS_LATEST = "tazrout/ai/decisions/latest";
    public static final String AI_ALERTS = "tazrout/ai/alerts";

    public static final String SYSTEM_GATEWAY_HEARTBEAT = "tazrout/system/gateway/heartbeat";
    public static final String SYSTEM_GATEWAY_STATUS = "tazrout/system/gateway/status";
    public static final String SYSTEM_CONTROL = "tazrout/system/control";
    public static final String SYSTEM_EMERGENCY_STOP = "tazrout/system/emergency/stop";
    public static final String SYSTEM_EMERGENCY_ALERT = "tazrout/system/emergency/alert";
    public static final String SYSTEM_EMERGENCY_STATUS = "tazrout/system/emergency/status";
    public static final String SYSTEM_BACKEND_STATUS = "tazrout/system/backend/status";

    public static final String DASHBOARD_SUMMARY = "tazrout/dashboard/summary";
    public static final String DASHBOARD_NOTIFICATIONS = "tazrout/dashboard/notifications";

    public static final String SETTINGS_PREFERENCES = "tazrout/settings/preferences";

    public static final String ANALYTICS_WATER_USAGE_PREFIX = "tazrout/analytics/water-usage/";
    public static final String ANALYTICS_ENV_PREFIX = "tazrout/analytics/environmental/";
    public static final String ANALYTICS_CONSUMPTION_PREFIX = "tazrout/analytics/consumption/";

    public static boolean isZoneSensorTopic(String topic) {
        return topic != null && topic.startsWith("tazrout/zones/") && topic.endsWith(ZONE_SENSORS_SUFFIX);
    }

    public static boolean isZoneStateTopic(String topic) {
        return topic != null && topic.startsWith("tazrout/zones/") && topic.endsWith(ZONE_STATE_SUFFIX);
    }

    public static boolean isZoneValveAckTopic(String topic) {
        return topic != null && topic.startsWith("tazrout/zones/") && topic.endsWith(ZONE_VALVE_ACK_SUFFIX);
    }
}

File: frontend/lib/core/constants/mqtt_topics.dart
//? Central registry of all MQTT topic strings used in the Flutter frontend.
//? Must stay in sync with docs/MQTT_TOPICS.md and backend MqttTopics.java.

//& MqttTopics Constants
abstract final class MqttTopics {
  //& Zone Topics
  static const String zonePrefix = 'tazrout/zones/';
  static const String zoneSensorsSuffix = '/sensors';
  static const String zoneStateSuffix = '/state';
  static const String zoneValveAckSuffix = '/valve/ack';

  //& AI Topics
  static const String aiDecisions = 'tazrout/ai/decisions';
  static const String aiDecisionsLatest = 'tazrout/ai/decisions/latest';
  static const String aiAlerts = 'tazrout/ai/alerts';

  //& Dashboard Topics
  static const String dashboardSummary = 'tazrout/dashboard/summary';
  static const String dashboardNotifications = 'tazrout/dashboard/notifications';

  //& System Topics
  static const String systemControl = 'tazrout/system/control';
  static const String systemEmergencyStop = 'tazrout/system/emergency/stop';
  static const String systemEmergencyStatus = 'tazrout/system/emergency/status';
  static const String systemEmergencyAlert = 'tazrout/system/emergency/alert';
  static const String systemGatewayHeartbeat = 'tazrout/system/gateway/heartbeat';
  static const String systemGatewayStatus = 'tazrout/system/gateway/status';
  static const String systemBackendStatus = 'tazrout/system/backend/status';

  //& Analytics Topics
  static const String analyticsWaterUsagePrefix = 'tazrout/analytics/water-usage/';
  static const String analyticsEnvPrefix = 'tazrout/analytics/environmental/';
  static const String analyticsConsumptionPrefix = 'tazrout/analytics/consumption/';

  //& Settings & Support
  static const String settingsPreferences = 'tazrout/settings/preferences';
  static const String supportContact = 'tazrout/support/contact';

  //& Topic Matchers
  static bool isZoneSensorTopic(String topic) =>
      topic.startsWith(zonePrefix) && topic.endsWith(zoneSensorsSuffix);

  static bool isZoneStateTopic(String topic) =>
      topic.startsWith(zonePrefix) && topic.endsWith(zoneStateSuffix);

  static String zoneIdFrom(String topic) {
    final parts = topic.split('/');
    return parts.length > 2 ? parts[2] : 'unknown';
  }
}
